FreeRTOS/Demo/CORTEX_R4_RM46_HERCULES_MCU_GCC/include/demo_tasks.h

```diff
@@ -0,0 +1,203 @@
+#define demoREG_UNPRIVILEGED_TASK_PRIORITY ( configMAX_PRIORITIES - 1UL )
+
+/** @brief Priority at which the prvQueueSendTask is created. */
+#define demoQUEUE_SEND_TASK_PRIORITY       ( tskIDLE_PRIORITY + 1UL )
+
+/** @brief Priority at which the prvQueueReceiveTask is created. */
+#define demoQUEUE_RECEIVE_TASK_PRIORITY    ( demoQUEUE_SEND_TASK_PRIORITY + 1UL )
+
+/** @brief Priority at which the MPU Read & Write Task is created. */
+#define demoMPU_READ_WRITE_TASK_PRIORITY   ( tskIDLE_PRIORITY + 3UL )
+
+/** @brief Priority at which the MPU Read Only Task is created. */
+#define demoMPU_READ_ONLY_TASK_PRIORITY    ( tskIDLE_PRIORITY + 4UL )
+
+/** @brief Priority at which the Nested IRQ Test Task is created. */
+#define demoIRQ_TASK_PRIORITY              ( configTIMER_TASK_PRIORITY + 2UL )
+
+/** @brief Priority at which the Notification Demo Task is created. */
+#define demoNOTIFICATION_TASK_PRIORITY     ( configTIMER_TASK_PRIORITY + 1UL )
+
+/* -------------------------- Register Test Tasks -------------------------- */
+
+/** @brief ASM function in reg_test_GCC.S that tests proper context swaps. */
+void vRegTest1Implementation( void );
+
+/** @brief ASM function in reg_test_GCC.S that tests proper context swaps. */
+void vRegTest2Implementation( void );
+
+/** @brief Creates the Register Test Tasks implemented in reg_test_GCC.S
+ * @return pdPASS if all tasks are created, pdFAIL if they are not.
+ */
+BaseType_t xCreateRegisterTestTasks( void );
+
+/* ------------------------ Demo Tasks Declarations ------------------------ */
+
+/** @brief Create two tasks, a queue, and a timer, which are used to blink an LED.
+ *
+ * @return
+ * pdPASS if all objects are created.
+ * pdFAIL if any object cannot be created.
+ */
+BaseType_t xCreateQueueTasks( void );
+
+/** @brief Create the MPU Tasks that trigger data aborts.
+ *
+ * @note The MPU demo creates 2 unprivileged tasks - One of which has Read Only
+ * access to a shared memory region while the other has Read Write access. The
+ * task with Read Only access then tries to write to the shared memory which
+ * results in a Memory fault. The fault handler examines that it is the fault
+ * generated by the task with Read Only access and if so, it recovers from the
+ * fault gracefully by moving the Program Counter to the next instruction to the
+ * one which generated the fault. If any other memory access violation occurs,
+ * the fault handler will get stuck in an infinite loop.
+ */
+BaseType_t xCreateMPUTasks( void );
+
+/** @brief Create a task that waits for a response from a nested IRQ
+ *
+ * @return pdPASS if tasks are created
+ * pdFAIl if tasks are not created
+ */
+BaseType_t xCreateIRQTestTask( void );
+
+/** @brief Create tasks that send task notifications back and forth.
+ *
+ * @return pdPASS if tasks are created
+ * pdFAIl if tasks are not created
+ */
+BaseType_t xCreateNotificationTestTask( void );
+
+/** @brief Interrupt Handler used for Software Raised Interrupts */
```

Comment:
This is actually correct since the `vIRQDemoHandler()` is only included, and called, if the source of the IRQ is the specific Software Raised Interrupt